totp/src/main/java/com/okta/totp/otp_display/OtpDisplayViewModel.kt

```diff
@@ -0,0 +1,134 @@
+
+@HiltViewModel
+class OtpDisplayViewModel @Inject constructor(
+    private val otpUriSharedPreferences: OtpUriSharedPreferences,
+    private val otpUriParser: OtpUriParser,
+    private val passwordGeneratorFactory: PasswordGeneratorFactory,
+    @IoDispatcher private val ioDispatcher: CoroutineDispatcher,
+) : ViewModel() {
+    private var limitUpdates = false
+    private var maxUpdates = 0
+    private val events = MutableSharedFlow<OtpDisplayViewModelEvents>()
+
+    private val otpEntryFlow = events.scan(getInitialOtpCodes()) { otpEntryList, event ->
+        when (event) {
+            is OtpDisplayViewModelEvents.UpdateOtpEntries -> {
+                otpEntryList.map { otpEntry ->
+                    val currentOtpCode = otpEntry.passwordGenerator.generate()
+                    otpEntry.copy(otpCode = currentOtpCode)
+                }
+            }
+            is OtpDisplayViewModelEvents.DeleteOtpViewmodelData -> {
+                otpUriSharedPreferences.removeOtpUriString(event.otpViewmodelData.otpUriString)
+                otpEntryList - event.otpViewmodelData
+            }
+        }
+    }
+        .flowOn(ioDispatcher)
+
+    val otpScreenUiStateFlow = otpEntryFlow.map { otpEntryList ->
+        otpEntryList.map { otpEntry ->
+            OtpEntry(
+                otpCode = otpEntry.otpCode,
+                account = otpEntry.account,
+                issuer = otpEntry.issuer,
+                onDeleteOtpEntry = {
+                    viewModelScope.launch(ioDispatcher) {
+                        events.emit(OtpDisplayViewModelEvents.DeleteOtpViewmodelData(otpEntry))
+                    }
+                }
+            )
+        }
+    }
+        .flowOn(ioDispatcher)
+
+    init {
+        viewModelScope.launch(ioDispatcher) {
+            var updates = 0
+            while (limitUpdates && updates++ < maxUpdates) {
+                delay(OTP_REFRESH_TIME_SECS.seconds)
+                events.emit(OtpDisplayViewModelEvents.UpdateOtpEntries)
+            }
+        }
+    }
+
+    private fun getInitialOtpCodes(): List<OtpViewmodelData> {
+        return otpUriSharedPreferences.getOtpUriStrings()
+            .map { otpUriString ->
+                val otpParams =
+                    otpUriParser.parseOtpUriString(otpUriString) as OtpUriParsingResults.OtpData
+                val passwordGenerator = passwordGeneratorFactory.getPasswordGenerator(otpParams)
+                val currentOtpCode = passwordGenerator.generate()
+                OtpViewmodelData(
+                    otpCode = currentOtpCode,
+                    account = otpParams.name,
+                    issuer = otpParams.issuer,
+                    passwordGenerator = passwordGenerator,
+                    otpUriString = otpUriString,
+                )
+            }
+    }
+
+    @VisibleForTesting
+    fun setMaxUpdateCycles(maxUpdates: Int) {
+        limitUpdates = true
+        this.maxUpdates = maxUpdates
+    }
```

Comment:
Had to add this so that init didn't run forever. I'm using runTest for OtpDisplayViewModelTest, and the test coroutine there automatically skips delays. I couldn't find a way to disable the delay-skipping behavior, so I've added this for now until I find a better way